modules/face/src/facemarkLBF.cpp

```diff
@@ -0,0 +1,1294 @@
+        }
+
+        std::vector<Rect> & faces = *(std::vector<Rect> *)roi.getObj();
+        if(faces.size()<1) return false;
+
+        std::vector<std::vector<Point2f> > & landmarks =
+            *(std::vector<std::vector<Point2f> >*) _landmarks.getObj();
+
+        landmarks.resize(faces.size());
+
+        for(unsigned i=0; i<faces.size();i++){
+            params.detectROI = faces[i];
+            fitImpl(image.getMat(), landmarks[i]);
+        }
+
+        return true;
+    }
+
+    bool FacemarkLBFImpl::fitImpl( const Mat image, std::vector<Point2f>& landmarks){
+        if (landmarks.size()>0)
+            landmarks.clear();
+
+        if (!isModelTrained) {
+           std::string error_message = "The LBF model is not trained yet. Please provide a trained model.";
+           CV_Error(CV_StsBadArg, error_message);
+        }
+
+        Mat img;
+        if(image.channels()>1){
+            cvtColor(image,img,CV_BGR2GRAY);
+        }else{
+            img = image;
+        }
+
+        Rect box;
+        if (params.detectROI.width>0){
+            box = params.detectROI;
+        }else{
+            std::vector<Rect> rects;
+
+            if(!isSetDetector){
+                defaultFaceDetector(img, rects);
+            }else{
+                faceDetector(img, rects,0);
+            }
+
+            if (rects.size() == 0)  return 0; //failed to get face
+            box = rects[0];
+        }
+
+        double min_x, min_y, max_x, max_y;
+        min_x = std::max(0., (double)box.x - box.width / 2);
+        max_x = std::min(img.cols - 1., (double)box.x+box.width + box.width / 2);
+        min_y = std::max(0., (double)box.y - box.height / 2);
+        max_y = std::min(img.rows - 1., (double)box.y + box.height + box.height / 2);
+
+        double w = max_x - min_x;
+        double h = max_y - min_y;
+
+        BBox bbox(box.x - min_x, box.y - min_y, box.width, box.height);
+        Mat crop = img(Rect((int)min_x, (int)min_y, (int)w, (int)h)).clone();
+        Mat shape = regressor.predict(crop, bbox);
+
+        if(params.detectROI.width>0){
+            landmarks = Mat(shape.reshape(2)+Scalar(min_x, min_y));
+            params.detectROI.width = -1;
+        }else{
+            landmarks = Mat(shape.reshape(2)+Scalar(min_x, min_y));
+        }
+
+        return 1;
+    }
+
+    void FacemarkLBFImpl::read( const cv::FileNode& fn ){
+        params.read( fn );
+    }
+
+    void FacemarkLBFImpl::write( cv::FileStorage& fs ) const {
+        params.write( fs );
+    }
+
+    void FacemarkLBFImpl::loadModel(String s){
+        if(params.verbose) printf("loading data from : %s\n", s.c_str());
+        std::ifstream infile;
+        infile.open(s.c_str(), std::ios::in);
+        if (!infile) {
+           std::string error_message = "No valid input file was given, please check the given filename.";
+           CV_Error(CV_StsBadArg, error_message);
+        }
+
+        FILE *fd = fopen(s.c_str(), "rb");
+        regressor.read(fd, params);
```

Comment:
OpenCV modules should not read/write files directly, especially via old "C-API" (there are a lot of missing error checks at least; there is no in-memory storage support; cross-platform issues).
Why FileStorage doesn't work here?

Reply:
ok, i will try to change to filestorage.

Reply:
Well, there is similar PR #1199 with same [problem](https://github.com/opencv/opencv_contrib/pull/1199/files#diff-bda6c2019eab834caf1e9453102e8025). Probably we should keep this as is if there is some problem with `cv::FileStorage`.
I believe it is fine for the first version. So consider this problem as minor too.